ocaml/xapi/import_raw_vdi.ml

```
@@ -98,32 +101,38 @@ let import vdi (req: Request.t) (s: Unix.file_descr) _ =
-            | Some sr ->
-              debug "Checking whether localhost can see SR: %s" (Ref.string_of sr);
-              if (Importexport.check_sr_availability ~__context sr)
-              then localhost_handler rpc session_id vdi req s
-              else
-                let host = Importexport.find_host_for_sr ~__context sr in
-                let address = Db.Host.get_address ~__context ~self:host in
-                return_302_redirect req s address;
+       try
+         Helpers.call_api_functions ~__context
+           (fun rpc session_id ->
+              let sr_opt = match vdi, sr_of_req ~__context req with
+                | Some vdi, _ -> Some (Db.VDI.get_SR ~__context ~self:vdi)
+                | None, Some sr -> Some sr
+                | None, None -> None
+              in
+              match sr_opt with
+              | Some sr ->
+                debug "Checking whether localhost can see SR: %s" (Ref.string_of sr);
+                if (Importexport.check_sr_availability ~__context sr)
+                then localhost_handler rpc session_id vdi req s  
+                else  
+                  let host = Importexport.find_host_for_sr ~__context sr in
+                  let address = Db.Host.get_address ~__context ~self:host in
+                  return_302_redirect req s address;
+                  None
+              | None ->
+                error "Require an SR or VDI to import";
+                let e = Api_errors.(Server_error(vdi_missing,[])) in 
+                fail_task_in_request req s e;
                 None
-            | None ->
-              error "Require an SR or VDI to import";
-              TaskHelper.failed ~__context Api_errors.(Server_error(vdi_missing,[]));
-              Http_svr.headers s (Http.http_400_badrequest ());
-              None
+           )
+       with e ->
+         error "Caught exception in import handler: %s" (ExnHelper.string_of_exn e);
+         fail_task_in_request req s e;
+         None
```

I think we should reraise the exception `e` here instead of returning `None`, to preserve the old behaviour for the other callers of this function.